lib/SILOptimizer/Differentiation/PullbackCloner.cpp

```diff
@@ -400,12 +422,61 @@ class PullbackCloner::Implementation final
     }
     /// If adjoint value is concrete, it is already materialized. Store it in
     /// the destination address.
-    case AdjointValueKind::Concrete:
+    case AdjointValueKind::Concrete: {
       auto concreteVal = val.getConcreteValue();
       builder.emitStoreValueOperation(loc, concreteVal, destAddress,
                                       StoreOwnershipQualifier::Init);
       break;
     }
+    case AdjointValueKind::AddElement: {
+      auto baseAdjoint = val;
+      auto baseAdjointType = baseAdjoint.getType();
+
+      // Current adjoint may be made up of layers of `AddElement` adjoints.
+      // We can iteratively gather the list of elements to add instead of making
+      // recursive calls to `materializeAdjointIndirect`.
+      SmallVector<AddElementValue *, 4> addEltAdjValues;
+
+      do {
+        auto addElementValue = baseAdjoint.getAddElementValue();
+        addEltAdjValues.push_back(addElementValue);
+        baseAdjoint = addElementValue->baseAdjoint;
+        assert(baseAdjointType == baseAdjoint.getType());
+      } while (baseAdjoint.getKind() == AdjointValueKind::AddElement);
```

Comment:
Do we have tests for this loop, btw? Maybe I'm missing something, but all loops are doing single struct/tuple extract?

Reply:
Not explicit tests but the other tests I have added do use this code path -- I needed to fix new test failures after making these changes. But I think it's a good idea to add explicit tests for this.

Reply:
@asl I have added tests exercising the nested add element adjoint materialization path.